specs/DesignDiscussions/DotNetTemplate.md

```diff
@@ -0,0 +1,348 @@
+- Data context is a data object adaptive element can use
+- Data context supports a memory concept, scope
+- Data context has two types, dictionary and array
+
+#### Scope
+Data objects can be placed anywhere in AdaptiveCards elements such as Container, TextBlock, and etc in json.
+Templating string is bound to the nearest data when AdaptiveCard is viewed as a tree
+
+```json
+{
+    "type": "AdaptiveCard",
+    "$data": {
+        "key": "1.2"
+    },
+    "body": [
+        {
+            "type": "Container",
+            "$data": {
+                "key": "2"
+            },
+            "items": [
+                {
+                    "type": "TextBlock",
+                    "$data": {
+                        "key": "3"
+                    },
+                    "text": "key should be 3, actual:{key}",
+                    "size": "Medium",
+                    "wrap": true
+                },
+                {
+                    "type": "TextBlock",
+                    "text": "key should be 2, actual:{key}",
+                    "size": "Medium",
+                    "wrap": true
+                }
+            ],
+            "style": "good",
+            "bleed": true
+        }
+    ],
+    "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+    "version": "{key}"
+}
+```
+
+The above json will be parsed as below image. First TextBlock will have "3", and the next "2"  
+  
+**Figure 4**
+![](antlr4_parse_tree_3.png)
+
+We also have **$root** key word. This is the data context at the root level meaning it will be set at AdaptiveCard level, and applies to all element with all the scoping rule respected as discussed
+
+### Type
+Data can be either dictionary type or array type
+
+```json
+"$data": 
+    { "name": "Matt" },
+```
+
+```json
+"$data": [
+    { "name": "Matt" },
+    { "name": "David" },
+    { "name": "Thomas" }
+],
+```
+### Parsing Strategy
+Since the introduction of a new data context happens only in a pair, and only an obj can include the pair, the Obj is the natural place to handle the data context. 
+In VisitObj method, it filters through the method's current context's children which are pairs, and tries to catch a data context node. 
+ Current design tentatively uses the first one it sees for the performance reason.  
+When a data context node is encountered, it call a base visit method with the node, then the method resolved into VisitTemplateData method.
+The method, creates a data context object with a field that indicates Array if set. The data context object parse the value in the pair of the data context node into JArray object if it's array, otherwise into JObject
+The newly created data context object is set to current context. In current design, it uses a stack for this purpose. When the method returns to the VisitObj method, data context is set, and translation can continue. Please refer to **Figure 4**
+* Do we use first data context we see, or do we have to see all of them?
```

Comment:
Discussion in the room: This is invalid json. We should do whatever we usually do with invalid json. (In shared model, for example, we don't render cards that are invalid json) 

Relatedly, as part of the inconsistency work we should confirm that we handle invalid json the same way on all platforms.